### PR TITLE
[serverless] Add EventBridge Inferred Spans

### PR DESCRIPTION
**Pull request by nhulston**
### What does this PR do?
- This PR fixes/adds EventBridge inferred spans for Lambda.
- Now, users will see an EventBride span on any Lambda function that was directly invoked by EventBridge.
- In the example architecture `Lambda -> EventBridge -> Lambda`, the traces look like this: Screenshot 2024-09-12 at 4 50 56 PM
> Two different traces. Second trace is missing an EventBridge span

- Now, the trace looks like this: Screenshot 2024-09-12 at 4 53 26 PM

### Motivation
SNS and SQS are already supported. There was some code in place for EventBridge, but it didn't work properly and didn't show spans on Datadog. This PR aims to fix this.

### Additional Notes

- I have a [PR here](https://github.com/DataDog/dd-trace-java/pull/7613) to inject trace context into the Java tracer. At the time of writing this, the tracers for .NET and Go do not inject trace context into SQS/SNS/EventBridge yet, but that's my next project.

- The AWS API has no reliable way of getting the start time of the EventBridge span. It only gives time with precision to the second, which is not precise enough to create spans. Therefore, we try to get the start time from the tracer in the field `x-datadog-sent-timestamp`. If this field does not exist, we fall back to AWS API's start time.

- Also, the AWS API has no way of getting the EventBridge bus name, so we attempt to get this from the tracer in the field `x-datadog-bus-name` for the span resource name. If this doesn't exist, we set the resource name to "EventBridge". 

- This should not be a breaking change, but this is an area for reviewers to focus on and double check. If the user is using an outdated tracer or .NET/Go that doesn't yet support inferred spans for EB/SQS/SNS, these changes shouldn't break anything, but it still shows an "EventBridge" span on any Lambda function invoked by the EventBridge event:
Screenshot 2024-09-16 at 4 29 37 PM

### Possible Drawbacks / Trade-offs
Increased overhead

### Describe how to test/QA your changes
Run local tests with `invoke test --targets=./pkg/serverless` from the root of the project

It's more complicated to test manually. Contact me if you want a live demo of these changes. Here's how I tested:
1. Create a serverless architecture that's supported, such as a Java Lambda function that publishes an event to EventBridge, an EventBridge event bus that has a target SQS queue or SNS notifier, and another Java Lambda function that subscribes to the queue/notifier.
2. Instrument the lambda functions using `datadog-ci` 
3. Build the extension: `./scripts/publish_sandbox.sh`
4. Build the new Java tracer from [this PR](https://github.com/DataDog/dd-trace-java/pull/7613): `./gradlew publishToMavenLocal && cd ~/.m2/repository/com/datadoghq/dd-java-agent/1.40.0-SNAPSHOT && zip layer.zip dd-java-agent-1.40.0-SNAPSHOT.jar`
5. Create new layers on AWS by uploading those zip files, and replace your Lambda layers with your custom layers.
6. Set the environment variable `JAVA_TOOL_OPTIONS: -javaagent:"/opt/dd-java-agent-1.40.0-SNAPSHOT.jar"` on your lambda functions to use the new Java tracer
7. Invoke your first lambda function, and check Datadog for results. The spans should be combined into a single trace.

### CODE REVIEW

File: pkg/serverless/trigger/events/events.go [hunk 1]
```diff
@@ -230,7 +230,10 @@ type KinesisRecord struct {
 type EventBridgeEvent struct {
 	DetailType string `json:"detail-type"`
 	Source     string
-	StartTime  string
+	Time       time.Time
+	Detail     struct {
+		TraceContext map[string]string `json:"_datadog"`
+	}
```
Comment (nhulston) [inline pkg/serverless/trigger/events/events.go:236]: This allows us to easily unmarshall the trace context while avoiding other fields in `detail` that the agent doesn't care about (i.e. the user's custom fields)

File: pkg/serverless/trace/propagation/carriers_test.go [hunk 2]
```diff
@@ -372,6 +372,60 @@ func TestSnsEntityCarrier(t *testing.T) {
...
+			},
+			expErr: "",
+		},
+		{
+			name: "missing_trace_context",
+			event: events.EventBridgeEvent{
+				Detail: struct {
+					TraceContext map[string]string `json:"_datadog"`
+				}{
+					TraceContext: map[string]string{},
```
Comment (purple4reina) [inline pkg/serverless/trace/propagation/carriers_test.go:408]: We should add a test for the `nil` case as well.

File: pkg/serverless/trace/propagation/extractor_test.go [hunk 3]
```diff
@@ -407,6 +407,48 @@ func TestExtractorExtract(t *testing.T) {
...
+			expNoErr: true,
+		},
+		{
+			name: "eventbridge-event-without-trace-context",
+			events: []interface{}{
+				events.EventBridgeEvent{
+					Detail: struct {
+						TraceContext map[string]string `json:"_datadog"`
+					}{
+						TraceContext: map[string]string{},
```
Comment (purple4reina) [inline pkg/serverless/trace/propagation/extractor_test.go:444]: We should test the `nil` case as well.

File: pkg/serverless/trace/propagation/extractor_test.go [hunk 4]
```diff
@@ -407,6 +407,48 @@ func TestExtractorExtract(t *testing.T) {
...
+			name: "eventbridge-event-with-trace-context",
+			events: []interface{}{
+				events.EventBridgeEvent{
+					Detail: struct {
+						TraceContext map[string]string `json:"_datadog"`
+					}{
+						TraceContext: map[string]string{
+							"x-datadog-trace-id":          "123456789",
+							"x-datadog-parent-id":         "987654321",
+							"x-datadog-sampling-priority": "1",
```
Comment (purple4reina) [inline pkg/serverless/trace/propagation/extractor_test.go:429]: Use the `headersDD` instead. You should also test the other header types `headersAll` and `headersW3C`. Take a look at the tests above and follow that pattern.

File: pkg/serverless/trace/propagation/extractor_test.go (hunk 3 above)
Comment (purple4reina) [reply pkg/serverless/trace/propagation/extractor_test.go:444]: Nevermind, you already added it! I just didn't see it.

File: pkg/serverless/trace/inferredspan/span_enrichment_test.go [hunk 5]
```diff
@@ -639,17 +639,39 @@ func TestEnrichInferredSpanWithEventBridgeEvent(t *testing.T) {
...
 }
 
+func TestEnrichInferredSpanWithEventBridgeEventNoBus(t *testing.T) {
+	var eventBridgeEvent events.EventBridgeEvent
+	_ = json.Unmarshal(getEventFromFile("eventbridge-no-bus.json"), &eventBridgeEvent)
+	inferredSpan := mockInferredSpan()
+	inferredSpan.EnrichInferredSpanWithEventBridgeEvent(eventBridgeEvent)
+	span := inferredSpan.Span
+	assert.Equal(t, uint64(7353030974370088224), span.TraceID)
+	assert.Equal(t, uint64(8048964810003407541), span.SpanID)
```
Comment (purple4reina) [inline pkg/serverless/trace/inferredspan/span_enrichment_test.go:660]: I don't think we neeed to test trace and span ids here, we just overwrite later anyway.

**Review by purple4reina — APPROVED** [review]

File: pkg/serverless/trace/inferredspan/constants.go [hunk 6]
```diff
@@ -11,9 +11,11 @@ const (
 	apiName          = "apiname"
 	bucketARN        = "bucket_arn"
 	bucketName       = "bucketname"
+	busName          = "x-datadog-bus-name"
```
Comment (duncanista) [inline pkg/serverless/trace/inferredspan/constants.go:14]: I'm not sure this tags should be here as this are like headers? 🤔  also, what about `x-datadog-resource-name`? i wonder if `bus-name` is going to be quite niche

File: pkg/serverless/trace/inferredspan/span_enrichment_test.go [hunk 7]
```diff
@@ -637,19 +637,35 @@ func TestEnrichInferredSpanWithEventBridgeEvent(t *testing.T) {
 	inferredSpan := mockInferredSpan()
 	inferredSpan.EnrichInferredSpanWithEventBridgeEvent(eventBridgeEvent)
 	span := inferredSpan.Span
-	assert.Equal(t, uint64(7353030974370088224), span.TraceID)
-	assert.Equal(t, uint64(8048964810003407541), span.SpanID)
```
Comment (duncanista) [inline pkg/serverless/trace/inferredspan/span_enrichment_test.go:641]: Why are these removed?

File: pkg/serverless/trace/propagation/carriers_test.go [hunk 8]
```diff
@@ -372,6 +372,72 @@ func TestSnsEntityCarrier(t *testing.T) {
...
+						"x-datadog-sampling-priority": "1",
+					},
+				},
+			},
+			expMap: map[string]string{
+				"x-datadog-trace-id":          "123456789",
+				"x-datadog-parent-id":         "987654321",
+				"x-datadog-sampling-priority": "1",
+			},
+			expErr: "",
```
Comment (duncanista) [inline pkg/serverless/trace/propagation/carriers_test.go:400]: add a w3c example

File: pkg/serverless/trace/propagation/extractor_test.go [hunk 9]
```diff
@@ -407,6 +407,72 @@ func TestExtractorExtract(t *testing.T) {
...
+		{
+			name: "eventbridge-event-with-w3c-headers",
+			events: []interface{}{
+				events.EventBridgeEvent{
+					Detail: struct {
+						TraceContext map[string]string `json:"_datadog"`
+					}{
+						TraceContext: headersMapW3C,
+					},
+				},
```
Comment (duncanista) [inline pkg/serverless/trace/propagation/extractor_test.go:456]: Ah, its here

File: pkg/serverless/trace/inferredspan/span_enrichment_test.go (hunk 7 above)
Comment (nhulston) [reply pkg/serverless/trace/inferredspan/span_enrichment_test.go:641]: Rey mentioned that we don't need to test them here because we overwrite them anyway

File: pkg/serverless/trace/inferredspan/constants.go (hunk 6 above)
Comment (nhulston) [reply pkg/serverless/trace/inferredspan/constants.go:14]: I noticed `sentTimestamp` was used by the SQS implementation, and it's located in this same file. I'm happy to move it if it doesn't make sense, but it is indeed a constant used for inferred spans, so I think it makes sense here. What do you think?

I agree that `x-datadog-resource-name` would be a better variable name! Gonna change that

File: pkg/serverless/trace/propagation/extractor_test.go [hunk 10]
```diff
@@ -407,6 +407,72 @@ func TestExtractorExtract(t *testing.T) {
...
+			},
+			expCtx:   w3cTraceContext,
+			expNoErr: true,
+		},
```
Comment (duncanista) [inline pkg/serverless/trace/propagation/extractor_test.go:460]: Could we add a W3C test with resource name + start time? Not sure where it should live, but I want to make sure we are covering that behavior.

**Review by duncanista — APPROVED** [review]
Will aprove, but we need another test to cover W3C + custom headers to see how it behaves. LMK if you can also test it e2e

File: pkg/serverless/trace/propagation/extractor_test.go (hunk 10 above)
Comment (nhulston) [reply pkg/serverless/trace/propagation/extractor_test.go:461]: Sure, I added a test in my [latest commit](https://github.com/DataDog/datadog-agent/pull/29414/commits/06c5ae4a0a28ccacba1336f15ea0c24bdca44e4e)
